src/s_inter.c

```diff
@@ -138,6 +203,514 @@ void sys_set_temppath(void);
+//
+// As a consequence of this:
+//  - for callbacks that are RT-friendly, sys_domicrosleep() does not need to select()
+//  - for callbacks that are not RT-friendly, sys_domicrosleep() doesn't have to select()
+//  at every call, but only when data is actually available.
+static void poll_fds()
+{
+    const unsigned int bufsize = pd_this->pd_inter->i_iothreadbufsize;
+    ssize_t ret;
+    struct timeval timout = {0}; // making this timeout longer would be more efficient (by avoiding spurious wakeups), but it would needlessly postpone writes
+    int pollem = 1;
+    int i;
+    t_fdpoll *fp;
+    if (pollem && pd_this->pd_inter->i_nfdpoll)
+    {
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerIoThread);
+        for (i = 0; i < pd_this->pd_inter->i_nfdpoll; i++)
+        {
+            t_fdpoll *fp = pd_this->pd_inter->i_fdpoll + i;
+            int fd = fp->fdp_fd;
+            if(fp->fdp_manager != kFdpManagerIoThread)
+                continue;
+            if(!FD_ISSET(fd, &readset))
+                continue;
+            if(fp->fdp_callback_in_audio_thread)
+            {
+                // temporarily transfer ownership
+                SYNC_STORE(&fp->fdp_manager, kFdpManagerAudioThread);
+            }
+            else
+            {
+                int fd = fp->fdp_fd;
+                t_rbskt* rbskt = fp->fdp_rbskt;
+                ring_buffer* rb = rbskt->rs_rb;
+                unsigned int size = rb_available_to_write(rb);
+                if(size > bufsize) {
+                    if(iothreadbuf_resize(size))
+                    {
+                        size = bufsize;
+                        if(rbskt->rs_preserve_boundaries)
+                            post("Incomplete packet received\n");
+                    }
+                }
+                char* buf = pd_this->pd_inter->i_iothreadbuf;
+                // we adapted socketreceiver_read and netsend_readbin to use
+                // rbskt_recv instead of recv. So here we are only reading from the
+                // socket and making the data available through rbskt_recv
+                ret = recv(fd, buf, size, 0);
+                if(ret < 0)
+                {
+                    size = 0;
+                    ret = -errno;
+                } else {
+                    size = ret;
+                }
+                // store the received data in the ringbuffer
+                if(rbskt->rs_preserve_boundaries)
+                    ret = rb_write_to_buffer(rb, 2, &ret, sizeof(ret), buf, size);
+                else {
+                    if(0 == ret) {
+                        // recv() returning 0 on a streaming socket means the
+                        // other hand has closed the connection
+                        rbskt->rs_closed = 1;
+                    } else if (ret < 0) {
+                        // an error has occurred, we store errno
+                        rbskt->rs_errno = -ret;
```

Comment:
`recv` can also fail with UDP sockets!

For example: https://stackoverflow.com/questions/34242622/windows-udp-sockets-recvfrom-fails-with-error-10054
(In Pd, we ignore 10054, see `socket_errno_udp`, but externals might not do this.)

Reply:
> recv can also fail with UDP sockets!

yes that's accounted for. The architecture for error reporting is as follows:
- if `rs_preserve_boundaries` is false (TCP), then the ring buffer only contains the data and `rs_errno` is set upon error.
- if `rs_preserve_boundaries` is true (UDP) then we send down the ring buffer the return value as well as the data.

When the `pollfn` calls `rbskt_recv()`, they will receive the value that was returned from `recv()`, transparently regardless of how the `rbskt` represents it internally.

Reply:
That is to say: the return value of `recv()` is parsed by the `pollfn()` as usual.